### PR TITLE
[Form] added prototype_data option in CollectionType

### PR DESCRIPTION
**Pull request by HeahDude**
| Q | A |
| --- | --- |
| Branch | 2.8+ |
| New doc | finishes #4367 |

### CODE REVIEW

File: reference/forms/types/collection.rst [hunk 1]
```diff
@@ -363,6 +364,29 @@ rendering your form, having the entire "form row" may be easier for you.
 For details on how to actually use this option, see the above example as
 well as :ref:`cookbook-form-collections-new-prototype`.
 
+prototype_data
+~~~~~~~~~~~~~~
+
+**type**: ``mixed`` **default**: ``null``
+
+Allows you to define a specific data for the prototype. Each new row added will
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: [...] define specific data [...]

**Comment by xabbuh** [comment]
We need a `versionadded` directive for this option.

File: reference/forms/types/collection.rst [hunk 2]
```diff
@@ -363,6 +364,29 @@ rendering your form, having the entire "form row" may be easier for you.
...
+By default, the same data as defined with ``entry_options`` for other entries
+will be used.
+
+.. code-block:: php
+
+    use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
+
+    $builder->add('tags', CollectionType::class, array(
+        'class' => \AppBundle\Entity\Tag::class,
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: For the Symfony 2.8 docs we always use a string to be compatible with PHP 5.3 and 5.4.

File: reference/forms/types/collection.rst [hunk 3]
```diff
@@ -363,6 +364,29 @@ rendering your form, having the entire "form row" may be easier for you.
...
 well as :ref:`cookbook-form-collections-new-prototype`.
 
+prototype_data
+~~~~~~~~~~~~~~
+
+**type**: ``mixed`` **default**: ``null``
+
+Allows you to define a specific data for the prototype. Each new row added will
+initially contain the data set by this option.
+By default, the same data as defined with ``entry_options`` for other entries
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: There's no need for the newline (to begin a new paragraph you would need a blank line).

File: reference/forms/types/collection.rst [hunk 4]
```diff
@@ -363,6 +364,29 @@ rendering your form, having the entire "form row" may be easier for you.
...
+will be used.
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: By default, the data configured for all entries with the `entry_options` will be used.

File: reference/forms/types/collection.rst (hunk 3 above)
Comment (HeahDude) [inline reference/forms/types/collection.rst]: Noted, thanks.

**Comment by HeahDude** [comment]
@xabbuh Thanks for the review, comments addressed.

File: reference/forms/types/collection.rst [hunk 5]
```diff
@@ -363,6 +364,31 @@ rendering your form, having the entire "form row" may be easier for you.
...
+**type**: ``mixed`` **default**: ``null``
+
+Allows you to define specific data for the prototype. Each new row added will
+initially contain the data set by this option. By default, the data configured
+for all entries with the ``entry_options`` option will be used.
+
+.. code-block:: php
+
+    use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: The use statements are useless now. :)

Comment (HeahDude) [inline reference/forms/types/collection.rst]: Right, I made another commit to do that change in all examples, so you can easily revert it in 3.0+ branches. Thanks!

**Comment by xabbuh** [comment]
:+1:

File: reference/forms/types/collection.rst [hunk 6]
```diff
@@ -53,13 +54,9 @@ in a form. For example, suppose you have an ``emails`` field that corresponds
 to an array of email addresses. In the form, you want to expose each email
 address as its own input text box::
 
-    use Symfony\Component\Form\Extension\Core\Type\CollectionType;
-    use Symfony\Component\Form\Extension\Core\Type\EmailType;
-    // ...
-
-    $builder->add('emails', CollectionType::class, array(
+    $builder->add('emails', 'Symfony\Component\Form\Extension\Core\Type\CollectionType', array(
```
Comment (weaverryan) [inline reference/forms/types/collection.rst]: Why the change away from the `CollectionType::class`? That option (with `::class`) *seems* a little easier because we can use IDE auto-completion. Or have we been changing this elsewhere?

Comment (HeahDude) [inline reference/forms/types/collection.rst]: @weaverryan, the discussion with @xabbuh is on the outdated diff: https://github.com/symfony/symfony-docs/pull/6450#discussion_r59172548
Because 2.8 supports PHP < 5.5, I made a different commit for those changes so it can be reverted in 3.0.
Thanks :)

Comment (xabbuh) [inline reference/forms/types/collection.rst]: @HeahDude It looks like I was mistaken here and we at some point agreed on using the `class` constant even in the 2.8 docs (see http://symfony.com/doc/2.8/book/forms.html, for example). Sorry for the confusion.

Comment (HeahDude) [inline reference/forms/types/collection.rst]: @xabbuh No worry, I just removed the last commit :)

File: reference/forms/types/collection.rst [hunk 7]
```diff
@@ -363,6 +364,32 @@ rendering your form, having the entire "form row" may be easier for you.
...
+for all entries with the ``entry_options`` option will be used.
+
+.. code-block:: php
+
+    use Symfony\Component\Form\Extension\Core\Type\CollectionType;
+    use Symfony\Component\Form\Extension\Core\Type\TextType;
+    // ...
+
+    $builder->add('tags', CollectionType::class, array(
+        'class' => 'AppBundle\Entity\Tag',
```
Comment (xabbuh) [inline reference/forms/types/collection.rst]: @HeahDude Can you revert the change here too? :)

Comment (HeahDude) [inline reference/forms/types/collection.rst]: @xabbuh, I intentionally didn't change this one since it's almost the same:
`'AppBundle\Entity\Tag'` VS `\AppBundle\Entity\Tag::class` unless we add a use statement:
```php
use AppBundle\Entity\Tag;

// ...
'class' => Tag::class
```
What do you like ?

Comment (xabbuh) [inline reference/forms/types/collection.rst]: Actually, I was a bit confused and I am even more confused now. Where from did you take the `class` option?

Comment (HeahDude) [inline reference/forms/types/collection.rst]: My bad, confusion with `EntityType`! Should be `entry_type` :)

Comment (xabbuh) [inline reference/forms/types/collection.rst]: Oh okay, but `entry_type` is already set to refer to the `TextType` isn't it?

**Comment by HeahDude** [comment]
Ok should be good now.

File: reference/forms/types/collection.rst (hunk 7 above)
Comment (HeahDude) [inline reference/forms/types/collection.rst]: Yes I answered too quickly and edited it.

**Comment by xabbuh** [comment]
Cool :+1:

**Comment by WouterJ** [comment]
Thanks for your work on this @HeahDude and @kgilden! It's finally merged now.

Fyi, I've added a reference to the entry_options setting in 326ae53.

**Comment by HeahDude** [comment]
Nice addition, thank you @WouterJ! Also thanks to @kgilden @xabbuh
